test/udruntime/include/updown.h

```diff
@@ -0,0 +1,1784 @@
+   *   |* * * * * * * |    | For expansion purposes
+   *   | * * * * * * *|    |
+   *   |--------------|  <-- NumLanes * CapacityControlPerLane
+   *   | * * * * * * *|
+   *   |* * * * * * * |   -|
+   *   | * * * * * * *|    | For expansion purposes
+   *   |* * * * * * * |   -|
+   *   | * * * * * * *|
+   *   |--------------|  <-- CapacityControlPerLane * CapacityNumLanes
+   * \endverbatim
+   *
+   */
+// retained in class
+// struct base_addr_t {
+//   volatile ptr_t mmaddr;
+//   volatile ptr_t spaddr;
+//   volatile ptr_t ctrlAddr;
+//   volatile ptr_t progAddr;
+// };
+
+/**
+ * @brief This is a class to manage the mapped memory
+ *
+ * This class cosiders memory allocation and deallocation
+ * it contains all the information needed to do this.
+ */
+class ud_mapped_memory_t {
+private:
+  /**
+     * @brief A region in memory.
+     *
+     * This struct represents a region in memory, full or empty
+     *
+     */
+  struct mem_region_t {
+    uint64_t size;  // Size of the region
+    bool     free;  // Is the region free or being used
+  };
+
+  // Map to keep track of the free and used regions.
+  //std::map<void *, mem_region_t> regions_local;  // local segments
+  //std::map<void *, mem_region_t> regions_global; // global segments area
+  std::map< void*,
+            mem_region_t,
+            std::less< void* >,
+            Allocator< std::pair< const void*, mem_region_t > > >
+    regions_local;  // local segments
+  std::map< void*,
+            mem_region_t,
+            std::less< void* >,
+            Allocator< std::pair< const void*, mem_region_t > > >
+    regions_global;  // global segments area
+
+public:
+  /**
+     * @brief Construct a new ud_mapped_memory_t object
+     *
+     * Initializes the regions with a single region
+     * that is free and contains all the elements
+     *
+     * @param machine information from the description of the current machine
+     */
+  ud_mapped_memory_t( ud_machine_t& machine ) {
+    UPDOWN_INFOMSG( "Initializing Mapped Memory Manager for %lu at 0x%lX",
+                    machine.MapMemSize,
+                    machine.MapMemBase );
+    // Create the first segment
+    void* baseAddr           = reinterpret_cast< void* >( machine.MapMemBase );
+    regions_local[baseAddr]  = { machine.MapMemSize, true };
+    baseAddr                 = reinterpret_cast< void* >( machine.GMapMemBase );
+    regions_global[baseAddr] = { machine.GMapMemSize, true };
+  }
+
+  /**
+     * @brief Get a new region in memory
+     *
+     * This is equivalent to malloc, it finds a free region that
+     * can allocate the current size, and, if there is free space
+     * left, it creates a new region with the remaining free space.
+     *
+     * @param size size to be allocated in bytes.
+     * @return void* pointer to the allocated memory
+     */
+  void* get_region( uint64_t size, bool global ) {
+    UPDOWN_INFOMSG( "Allocating new region of size %lu bytes", size );
```

Comment:
It might be better to use `"%" PRIu64` instead of `%lu`. `PRIu64` and similar macros are defined in `<cinttypes>` (C++) or `<inttypes.h>` (C). We have warnings as errors when `printf()` arguments don't match their format's size.